builtin/plugins/coin/coin.go

```diff
@@ -209,10 +254,223 @@ func mint(ctx contract.Context, to loom.Address, amount *loom.BigUInt) error {
+		if modulus.Cmp(big.NewInt(1)) == 0 || err == contract.ErrNotFound {
+			//Minting Amount Computation for starting year
+			amount, err = ComputeforFirstYear(ctx, baseAmount, blocksGeneratedPerYear)
+			if err != nil {
+				return errUtil.Wrap(err, "Failed Minting Block for first year")
+			}
+		}
+
+	} else {
+		if modulus.Cmp(big.NewInt(1)) == 0 && err != contract.ErrNotFound {
+			//Operator Support to support different inflation ratio patterns for different year
+			if strings.EqualFold("div", operator) {
+				amount, err = ComputeforConsecutiveYearBeginningDivOperator(ctx, baseAmount, changeRatioNumerator,
+					changeRatioDenominator, basePercentage, blocksGeneratedPerYear, amount, loom.NewBigUIntFromInt(relativeYear.Int64()))
+				if err != nil {
+					return errUtil.Wrap(err, "Failed Minting Block with div operator for consecutive year")
+				}
+			} else if strings.EqualFold("exp", operator) {
+				amount, err = ComputeforConsecutiveYearBeginningExpOperator(ctx, baseAmount, changeRatioNumerator,
+					changeRatioDenominator, basePercentage, blocksGeneratedPerYear, amount, loom.NewBigUIntFromInt(relativeYear.Int64()))
+				if err != nil {
+					return errUtil.Wrap(err, "Failed Minting Block Configuration with exp operator for consecutive year")
+				}
+			} else {
+				return errors.New("Invalid operator - Operator should be div or exp")
+			}
+		} else if err == contract.ErrNotFound {
+			amount, err = ComputeforFirstYearBlockHeightgreaterthanOneyear(ctx, baseAmount, basePercentage, blocksGeneratedPerYear)
+			if err != nil {
+				return errUtil.Wrap(err, "Failed Minting Block Configuration for exp operator")
+			}
+		} else {
+			amount, err = ComputeforConsecutiveYearinMiddle(ctx)
+
+		}
+	}
+	var mintingAmount = &types.BigUInt{
+		Value: *loom.NewBigUIntFromInt(0),
+	}
+	err = ctx.Get(mintingAmountKey, mintingAmount)
+	if err != nil {
+		return err
+	}
+	//Boundary case when minting amount per block becomes zero
+	if mintingAmount.Value.Cmp(loom.NewBigUIntFromInt(0)) == 0 {
+		return nil // No more coins to be minted on block creation
+	}
+	return mint(ctx, loom.UnmarshalAddressPB(policy.MintingAccount), loom.NewBigUIntFromInt(mintingAmount.Value.Int64()))
+}
+
+//First year is started from the checkpoint at which minting started
+//Computes minting Amount for first year if Blocks generated are less than a year, whether its year beginning or in the middle
+//Divides BaseAmountforfirstyear/Blocks generated per year to get minting Amount per block
+func ComputeforFirstYear(ctx contract.Context, baseAmount *common.BigUInt,
+	blocksGeneratedPerYear *common.BigUInt) (*common.BigUInt, error) {
+	amount := baseAmount.Div(baseAmount, blocksGeneratedPerYear)
+	err := ctx.Set(mintingAmountKey, &types.BigUInt{
+		Value: *amount,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return amount, nil
+}
+
+//Computes minting Amount for first year if Block Height is greater than blocks generated in a year, whether its year beginning or in the middle
```

Comment:
Why would we need to measure if Block height is greater than blocks generated in a year.

Reply:
Also, keep in mind that, we will start counting year from the height plugin is enabled.

Reply:
Algorithm Description for calculate base minting Amount for first year.

Two cases ==>

Case 1
Minting starts when first block of blockchain is produced.
In this case there will not be any total supply.
So, minting Amount per block will be computed as Base Amount for first year/Blocks generated in a year

Case 2
Minting starts at a height which is greater than blocks generated in a year.
If block height is greater than one year, when minting starts.
In this case, there will be some total supply accumulated for running period of blockchain, so minting amount per block will be computed as follows -
Base percentage for first year will be applied on top of total supply to get inflation for first year.
This inflation computed will be divided by Blocks generated in a year to get minting amount per block.

Reply:
Have taken that into account already, you can see there is a minting height which is computed when minting happens for the first time in blockchain, rest of the computations and year computations are relative to this.
```
if err == contract.ErrNotFound {
    //Sets Minting Height - Height at which minting started
        err := ctx.Set(mintingHeightKey, &types.BigUInt{
            Value: *loom.NewBigUIntFromInt(ctx.Block().Height),
        })
        if err != nil {
            return err
        }
    }
    var mintingHeight = &types.BigUInt{
        Value: *loom.NewBigUIntFromInt(0),
    }
    err1 := ctx.Get(mintingHeightKey, mintingHeight)
    if err1 != nil {
        return err1
    }
   
```

Above is code for the same

```
relativeHeight := big.NewInt(ctx.Block().Height).Sub(big.NewInt(ctx.Block().Height), big.NewInt(mintingHeight.Value.Int64()))
```

```
relativeYear := relativeHeight.Div(relativeHeight, blocksGeneratedPerYear.Int)
```

relativeHeight and relativeYear computation